src/libos/src/time/timer_file.rs

```diff
@@ -0,0 +1,394 @@
+                .task()
+                .tirqs()
+                .put_req(0);
+        }
+    }
+}
+
+impl TimerFile {
+    pub fn new(clockid: ClockId, flags: TimerCreationFlags) -> Result<Self> {
+        Ok(Self {
+            clockid,
+            flags: Atomic::new(flags),
+            inner: Arc::new(Mutex::new(TimerInner::new())),
+            pollee: Pollee::new(Events::empty()),
+        })
+    }
+
+    // The inner implementation of syscall timerfd_settime.
+    pub fn set_time(
+        &self,
+        flags: TimerSetFlags,
+        new_value: &TimerfileDurations,
+    ) -> Result<TimerfileDurations> {
+        let cur_time =
+            timespec_t::from(vdso_time::clock_gettime(self.clockid).unwrap()).as_duration();
+
+        let inner_c = self.inner.clone();
+        let mut inner = inner_c.lock().unwrap();
+        let ret = TimerfileDurations {
+            it_interval: inner.interval,
+            it_value: inner.init_exp,
+        };
+
+        // Check if the it_value is 0 which means stop the timer
+        if new_value.it_value.is_zero() == true {
+            debug!("TimerFd: stop timer");
+            inner.status = TimerFdStatus::STOP;
+            if inner.task_handle.is_some() {
+                // If timer task started, send signal to end the task
+                inner
+                    .task_handle
+                    .as_ref()
+                    .unwrap()
+                    .task()
+                    .tirqs()
+                    .put_req(0);
+            }
+
+            inner.task_handle = None;
+
+            return Ok(ret);
+        }
+
+        // Transfer the initial expired time to absolute time
+        let exp_time = match flags {
+            TimerSetFlags::TFD_TIMER_ABSTIME => new_value.it_value,
+            _ => cur_time.checked_add(new_value.it_value).unwrap(),
+        };
+
+        // Transfer the initial expired time to relative duration time.
+        // The relative time it_value will be used in wait_timeout.
+        let it_value = match exp_time.checked_sub(cur_time) {
+            Some(duration) => duration,
+            _ => Duration::new(0, 0),
+        };
+
+        inner.interval = new_value.it_interval;
+        inner.init_exp = exp_time;
+        inner.exp_cnt = 0;
+        inner.last_exp_cnt = 0;
+
+        if new_value.it_interval.is_zero() == true {
+            inner.status = TimerFdStatus::ONE_SHOT;
+        } else {
+            inner.status = TimerFdStatus::LOOP;
+        }
+
+        // If no timer task started, start it
+        if inner.task_handle.is_none() {
+            // Drop Mutex inner here.
+            // Will acquire it in fn self.start_timer_task again.
+            drop(inner);
+            self.start_timer_task(it_value);
+        }
+
+        return Ok(ret);
+    }
+
+    // The inner implementation of syscall timerfd_gettime.
+    pub fn time(&self) -> Result<TimerfileDurations> {
+        let mut ret_time = TimerfileDurations::default();
```

Comment:
Maybe we could implement this function in inner. And TimerFile could simply invoke the  inner time function.

Reply:
The inner is protected by SgxMutex. 
Put the function in inner implies the funtion will hold the lock all the time which is not efficient.